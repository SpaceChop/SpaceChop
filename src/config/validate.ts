import console from '../lib/console';
import validate from '../types/validate';

let valid = true;

export default (config, throwError = false) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.info('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.error('\nThere is an error in /config.yml:');
    console.error('--------------------------------');
    console.error(err.message);
    console.error('--------------------------------');
    if (throwError) {
      throw new Error(err);
    }
  }
  return false;
};
