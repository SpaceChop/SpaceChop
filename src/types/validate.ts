import Config from './Config';

export default (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key ? `${err.key}: ` : ''}${err.message}`);
  }
};
