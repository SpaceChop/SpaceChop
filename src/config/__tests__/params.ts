import validate from '../../types/validate';

describe('validate params', () => {
  describe('storage', () => {
    it('should throw error on missing param in storage', () => {
      const config = {
        paths: ['/:preset/:uuid/:image'],
        sources: [{
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: ':uuid',
          },
        }],
        storage: {
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: '/:preset',
          },
        },
        presets: {},
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });

    it('should throw error on missing param in path from storage', () => {
      const config = {
        paths: ['/:preset/:uuid'],
        sources: [{
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: ':uuid',
          },
        }],
        storage: {
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: '/:preset/:image',
          },
        },
        presets: {},
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });

    it('should not throw error on valid storage', () => {
      const config = {
        paths: ['/:preset/:image'],
        sources: [{
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: ':image',
          },
        }],
        storage: {
          s3: {
            access_key_id: 'access_key_id',
            secret_access_key: 'secret_access_key',
            region: 'region',
            bucket_name: 'bucket_name',
            path: ':image',
          },
        },
        presets: {},
      };
      expect(() => validate(config)).not.toThrow();
    });
  });

  describe('sources', () => {
    it('should throw on missing params in source', () => {
      const config = {
        paths: ['/:preset/:image'],
        sources: [{ s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: 'images',
        },
      }], presets: {} };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });

    it('should throw on missing params in path on source', () => {
      const config = {
        paths: ['/:preset/:uuid'],
        sources: [{ s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: 'images/:image',
        },
      }], presets: {} };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('presets', () => {
    it('should throw on missing path params in config', () => {
      let config = {
        paths: ['/:preset/:width/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: {},
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset/:width/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: 200,
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset/:width/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: 200,
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset/:width/:image'],
        sources: [{ volume: { root: '/src/app/test/assets' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: {
                from_path: 'width',
              },
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset/:width/:height/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: {
                from_path: 'width',
              },
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: {
                from_path: 'width',
              },
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
      config = {
        paths: ['/:preset'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: 200,
              height: 200,
            },
          }],
        } },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });

    it('should not throw on valid params in path', () => {
      const config = {
        paths: ['/:preset/:width/:height/:image'],
        sources: [{ volume: { root: '/src/app/test/assets/:image' } }],
        presets: { t_200: {
          steps: [{
            $resize: {
              width: {
                from_path: 'width',
              },
              height: {
                from_path: 'height',
              },
            },
          }],
        } },
      };
      expect(() => validate(config)).not.toThrow();
    });
  });
});
