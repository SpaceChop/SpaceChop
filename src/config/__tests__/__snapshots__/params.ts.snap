// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate params presets should throw on missing path params in config 1`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate params presets should throw on missing path params in config 2`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate params presets should throw on missing path params in config 3`] = `"params.width: Could not find any usage in sources, storage or paths"`;

exports[`validate params presets should throw on missing path params in config 4`] = `"params.image: Could not find any usage in sources, storage or paths"`;

exports[`validate params presets should throw on missing path params in config 5`] = `"params.height: Could not find any usage in sources, storage or paths"`;

exports[`validate params presets should throw on missing path params in config 6`] = `"presets.t_200.steps[0].width: Missing param (width) in paths for preset (t_200)"`;

exports[`validate params presets should throw on missing path params in config 7`] = `"source[0].volume.root: Missing param \\"image\\" in paths"`;

exports[`validate params sources should throw on missing params in path on source 1`] = `
"source[0].s3.path: Missing param \\"image\\" in paths
params.uuid: Could not find any usage in sources, storage or paths"
`;

exports[`validate params sources should throw on missing params in source 1`] = `
"storage.s3.path: Missing path parameters
params.image: Could not find any usage in sources, storage or paths"
`;

exports[`validate params storage should throw error on missing param in path from storage 1`] = `"storage.s3.path: Missing param \\"image\\" in paths"`;

exports[`validate params storage should throw error on missing param in storage 1`] = `
"storage.s3.path: Missing path parameters
params.image: Could not find any usage in sources, storage or paths"
`;
